rfcs/0100-sign-commits.md

```diff
@@ -0,0 +1,380 @@
+  make sure the version of Nixpkgs they are using is trusted.
+
+# Motivation
+[motivation]: #motivation
+
+There is already much literature on why you should
+sign your commits, but to quickly reiterate,
+it provides assurance that you can trust a commit,
+and thus makes it harder to inject malicious code
+during an update, i.e. moving from an old commit
+to a new commit.
+
+However, even if you sign your commits, Nix currently
+provides no way of automatically verifying said
+commits for trust.
+
+[Guix](https://guix.gnu.org/), a competitor to Nix, already
+support verifying commits through their own system, described
+in the following links:
+- **https://guix.gnu.org/en/blog/2020/securing-updates/**
+- https://github.com/guix-mirror/guix/commits/master
+- https://guix.gnu.org/manual/en/html_node/Commit-Access.html
+- https://guix.gnu.org/manual/en/html_node/Channel-Authentication.html
+
+In addition, most conventional Linux distributions already sign their
+binary packages. By adding this functionality to Nix, end-users of Nixpkgs
+and other Nix repositories can be sure that they're getting the correct
+source code, **regardless** of the channel through they received
+it.
+
+This RFC was motivated by work during Summer of Nix, specifically,
+the Libre SoC maintainer requested that source integrity be
+maintained.
+
+# Detailed design
+[design]: #detailed-design
+
+## Verifying trust
+[verifying trust]: #verifying-trust
+
+### The algorithm
+
+The design is inspired by what is described here: https://guix.gnu.org/en/blog/2020/securing-updates/.
+
+We have two inputs to consider, an old trusted commit,
+and a new potentially untrusted commit.
+
+In the context of Git, a parent is a direct parent in the DAG of commits.
+
+The old commit is always trusted.
+
+If a commit has a trusted parent, then the commit is trusted if
+it doesn't have a trusted parent with `/.well-known/authorized-committers.nix`,
+
+If a commit has a trusted parent,
+and is signed by a key listed in each of the trusted
+parents' `/.well-known/authorized-committers.nix`,
+and the key hasn't expired or been revoked, regardless
+of when the commit has made, then the commit is trusted.
+
+The reason the timestamp of the commit doesn't matter,
+is that it's not secure and can be forged.
+
+The reason that a commit only needs one trusted parent,
+instead of all parents being trusted like Guix, is that
+this allows merging non-trusted commits, while the merge
+commit itself would be trusted.
+
+### The format of `/.well-known/authorized-committers.nix`
+
+An example: `/.well-known/authorized-committers.nix`:
+```nix
+{
+  committers = {
+    jacob = "XXXX XXXX XXXX XXXX XXXX  XXXX XXXX XXXX XXXX XXXX";
+    jonathan = "YYYY YYYY YYYY YYYY YYYY  YYYY YYYY YYYY YYYY YYYY";
+  };
+}
+```
+
+Specifically, it will contain a Nix a set with a single attribute `committers`, which contains
+another set where names are mapped to PGP fingerprints. This is a set
```

Comment:
Does gpg not have the issue that it can expire? What happen if someone tries to build an old repository?

Reply:
All this logic is only relevant in the case of updating the commit used. If you specify a commit that's what you're going to get.
If you're trying to update from a repository with expired keys, then of course as you'd expect you do not get the same guarantees. How it will be handled depends on the exact implementation, you could e.g. simply print a warning and go on with the algorithm as usual, or you could err and require that the user manually find a new commit without expired keys.
I don't think the exact design is that important, since this is very much an edge case.